File: scripts/analyze_aiff.py
#!/usr/bin/env python3
"""
AIFF/AIFC File Analyzer

This script analyzes and compares AIFF/AIFC files to help debug OP-Z compatibility issues.
Usage: python3 analyze_aiff.py path/to/file1.aif [path/to/file2.aif]
"""

import os
import sys
import json
import struct
from pathlib import Path


def format_size(bytes_):
    """Format file size in human-readable form"""
    if bytes_ < 1024:
        return f"{bytes_} bytes"
    elif bytes_ < 1024 * 1024:
        return f"{bytes_ / 1024:.2f} KB"
    return f"{bytes_ / (1024 * 1024):.2f} MB"


def detect_file_type(data):
    """Detect if file is AIFF or AIFC"""
    if len(data) < 12:
        return "Unknown (too small)"
    
    form_id = data[0:4].decode('ascii', errors='replace')
    if form_id != 'FORM':
        return "Not an AIFF/AIFC file"
    
    form_type = data[8:12].decode('ascii', errors='replace')
    if form_type == 'AIFF':
        return "AIFF"
    elif form_type == 'AIFC':
        return "AIFC"
    
    return f"Unknown IFF type: {form_type}"


def parse_form_chunk(data):
    """Parse the main FORM chunk"""
    size = struct.unpack('>I', data[4:8])[0]
    form_type = data[8:12].decode('ascii', errors='replace')
    
    return {
        'id': 'FORM',
        'size': size,
        'form_type': form_type
    }


def parse_chunks(data):
    """Parse all chunks in the file"""
    chunks = []
    pos = 12  # Skip FORM header
    
    while pos + 8 <= len(data):
        id_ = data[pos:pos+4].decode('ascii', errors='replace')
        size = struct.unpack('>I', data[pos+4:pos+8])[0]
        
        chunks.append({
            'id': id_,
            'size': size,
            'offset': pos,
            'data': data[pos+8:pos+8+size]
        })
        
        pos += 8 + size
        # Add pad byte if chunk size is odd
        if size % 2 == 1:
            pos += 1
    
    return chunks


def analyze_aiff_file(file_path):
    """Analyze an AIFF or AIFC file"""
    print(f"\n=== Analyzing {Path(file_path).name} ===")
    
    try:
        with open(file_path, 'rb') as f:
            data = f.read()
        
        file_type = detect_file_type(data)
        print(f"File type: {file_type}")
        print(f"File size: {format_size(len(data))}")
        
        # Parse main FORM chunk
        form_chunk = parse_form_chunk(data)
        print(f"Form type: {form_chunk['form_type']}")
        print(f"Form size: {format_size(form_chunk['size'])}")
        
        # Parse all chunks
        chunks = parse_chunks(data)
        print("\nChunk structure:")
        for i, chunk in enumerate(chunks):
            print(f"{i + 1}. {chunk['id'].ljust(4)} | Size: {format_size(chunk['size']).ljust(10)} | Offset: 0x{chunk['offset']:08x}")
            
            # Special handling for known chunk types
            if chunk['id'] == 'FVER':
                version = struct.unpack('>I', chunk['data'][0:4])[0]
                print(f"   Version: 0x{version:x} ({version})")
            
            elif chunk['id'] == 'COMM':
                num_channels = struct.unpack('>H', chunk['data'][0:2])[0]
                num_frames = struct.unpack('>I', chunk['data'][2:6])[0]
                sample_size = struct.unpack('>H', chunk['data'][6:8])[0]
                print(f"   Channels: {num_channels}, Frames: {num_frames}, Bits: {sample_size}")
                
                if form_chunk['form_type'] == 'AIFC' and len(chunk['data']) >= 22:
                    comp_type = chunk['data'][18:22].decode('ascii', errors='replace')
                    print(f"   Compression: '{comp_type}'")
            
            elif chunk['id'] == 'APPL':
                if len(chunk['data']) >= 4:
                    app_type = chunk['data'][0:4].decode('ascii', errors='replace')
                    print(f"   App type: '{app_type}'")
                    
                    if app_type == 'op-1':
                        # This is OP-1 metadata, try to parse as JSON
                        try:
                            json_str = chunk['data'][4:].decode('utf-8', errors='replace')
                            metadata = json.loads(json_str)
                            
                            print(f"   Drum version: {metadata.get('drum_version', 'N/A')}")
                            print(f"   Name: {metadata.get('name', 'N/A')}")
                            
                            if 'start' in metadata and len(metadata['start']) > 0:
                                print(f"   Slices: {len(metadata['start'])}")
                                
                                # Calculate slice spacing to check for overlap
                                slice_info = []
                                for i, (start, end) in enumerate(zip(metadata.get('start', []), metadata.get('end', []))):
                                    if not start and not end:
                                        continue
                                    
                                    start_frames = round(start / 4096)
                                    end_frames = round(end / 4096)
                                    duration = (end_frames - start_frames) / 44100
                                    
                                    slice_info.append({
                                        'index': i,
                                        'start': start_frames,
                                        'end': end_frames,
                                        'duration': f"{duration:.3f}"
                                    })
                                
                                # Print slice information
                                print(f"\n   Slice details (first 10):")
                                for i, slice_ in enumerate(slice_info[:10]):
                                    if slice_['start'] == slice_['end']:
                                        continue  # Skip empty slices
                                    print(f"     Slice {slice_['index']}: Start={slice_['start']}, End={slice_['end']}, Duration={slice_['duration']}s")
                                
                                # Check for overlapping slices
                                has_overlap = False
                                for i in range(len(slice_info) - 1):
                                    current = slice_info[i]
                                    next_ = slice_info[i + 1]
                                    if current['end'] >= next_['start']:
                                        print(f"   WARNING: Overlap between slices {current['index']} and {next_['index']}")
                                        has_overlap = True
                                
                                if not has_overlap:
                                    print("   No slice overlaps detected")
                        
                        except json.JSONDecodeError as e:
                            print(f"   Error parsing metadata: {e}")
                        except Exception as e:
                            print(f"   Error processing metadata: {e}")
        
        # Look for specific issues based on our previous debugging
        print("\nAnalyzing for known issues:")
        
        # Check FVER chunk in AIFC files
        if form_chunk['form_type'] == 'AIFC':
            fver_chunks = [c for c in chunks if c['id'] == 'FVER']
            if not fver_chunks:
                print("❌ ERROR: AIFC file missing required FVER chunk")
            else:
                fver = fver_chunks[0]
                fver_pos = chunks.index(fver)
                if fver_pos != 0:
                    print(f"❌ WARNING: FVER chunk should be the first chunk after FORM, found at position {fver_pos + 1}")
                
                version = struct.unpack('>I', fver['data'][0:4])[0]
                if version != 0xa2805140:
                    print(f"❌ WARNING: FVER version 0x{version:x} doesn't match TE standard 0xa2805140")
                else:
                    print("✅ FVER chunk version matches TE standard")
        
        # Check COMM chunk
        comm_chunks = [c for c in chunks if c['id'] == 'COMM']
        if not comm_chunks:
            print("❌ ERROR: Missing required COMM chunk")
        else:
            comm = comm_chunks[0]
            if form_chunk['form_type'] == 'AIFC':
                if len(comm['data']) < 22:
                    print("❌ ERROR: AIFC COMM chunk too short, missing compression type")
                else:
                    comp_type = comm['data'][18:22].decode('ascii', errors='replace')
                    if comp_type != 'sowt':
                        print(f"❌ WARNING: COMM compression '{comp_type}' doesn't match TE standard 'sowt'")
                    else:
                        print("✅ COMM chunk has correct compression type 'sowt'")
        
        # Check APPL chunk
        appl_chunks = [c for c in chunks if c['id'] == 'APPL']
        if not appl_chunks:
            print("❌ WARNING: Missing APPL chunk with OP-1 metadata")
        else:
            appl = appl_chunks[0]
            if len(appl['data']) < 4 or appl['data'][0:4].decode('ascii', errors='replace') != 'op-1':
                print("❌ WARNING: APPL chunk doesn't have 'op-1' identifier")
            else:
                try:
                    json_str = appl['data'][4:].decode('utf-8', errors='replace')
                    metadata = json.loads(json_str)
                    
                    # Check drum_version
                    drum_version = metadata.get('drum_version')
                    if drum_version is None:
                        print("❌ ERROR: Missing drum_version in metadata")
                    elif (form_chunk['form_type'] == 'AIFC' and drum_version != 2) or \
                         (form_chunk['form_type'] == 'AIFF' and drum_version != 3):
                        print(f"❌ WARNING: drum_version {drum_version} doesn't match expected value for {form_chunk['form_type']}")
                    else:
                        print(f"✅ drum_version {drum_version} matches expected value for {form_chunk['form_type']}")
                    
                    # Check playmode values
                    playmode = metadata.get('playmode', [])
                    if not playmode:
                        print("❌ ERROR: Missing playmode values in metadata")
                    elif form_chunk['form_type'] == 'AIFC' and any(v != 4096 for v in playmode if v):
                        print(f"❌ WARNING: playmode values should be 4096 for AIFC")
                    elif form_chunk['form_type'] == 'AIFF' and any(v != 8192 for v in playmode if v):
                        print(f"❌ WARNING: playmode values should be 8192 for AIFF")
                    else:
                        print(f"✅ playmode values match expected values for {form_chunk['form_type']}")
                
                except Exception as e:
                    print(f"❌ ERROR: Failed to parse APPL metadata: {e}")
        
        return {
            'file_type': file_type,
            'form_type': form_chunk['form_type'],
            'chunks': chunks
        }
    
    except Exception as e:
        print(f"❌ ERROR: {e}")
        return None


def compare_files(file1, file2):
    """Compare two AIFF/AIFC files"""
    print(f"\n=== Comparing {Path(file1).name} vs {Path(file2).name} ===")
    
    data1 = analyze_aiff_file(file1)
    data2 = analyze_aiff_file(file2)
    
    if not data1 or not data2:
        print("Cannot compare - one or both files failed analysis")
        return
    
    print("\n=== Key Differences ===")
    
    # Compare file types
    if data1['file_type'] != data2['file_type']:
        print(f"❌ File type: {data1['file_type']} vs {data2['file_type']}")
    
    if data1['form_type'] != data2['form_type']:
        print(f"❌ Form type: {data1['form_type']} vs {data2['form_type']}")
    
    # Compare chunk structure
    chunks1 = [c['id'] for c in data1['chunks']]
    chunks2 = [c['id'] for c in data2['chunks']]
    
    print("\nChunk structure:")
    print(f"File 1: {', '.join(chunks1)}")
    print(f"File 2: {', '.join(chunks2)}")
    
    # Find missing chunks
    missing_in_1 = set(chunks2) - set(chunks1)
    missing_in_2 = set(chunks1) - set(chunks2)
    
    if missing_in_1:
        print(f"❌ Missing in File 1: {', '.join(missing_in_1)}")
    
    if missing_in_2:
        print(f"❌ Missing in File 2: {', '.join(missing_in_2)}")


# Main execution
if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Usage: python3 analyze_aiff.py path/to/file.aif [path/to/another-file.aif]")
    elif len(sys.argv) == 2:
        analyze_aiff_file(sys.argv[1])
    else:
        compare_files(sys.argv[1], sys.argv[2])

File: scripts/analyze-aiff.js
/**
 * AIFF/AIFC File Analyzer
 * 
 * This script analyzes and compares AIFF/AIFC files to help debug OP-Z compatibility issues.
 * Usage: node analyze-aiff.js path/to/file1.aif [path/to/file2.aif]
 */

const fs = require('fs');
const path = require('path');

// Main function
function analyzeAiffFile(filePath) {
  console.log(`\n=== Analyzing ${path.basename(filePath)} ===`);
  
  try {
    const buffer = fs.readFileSync(filePath);
    const fileType = detectFileType(buffer);
    console.log(`File type: ${fileType}`);
    console.log(`File size: ${formatSize(buffer.length)}`);
    
    // Parse main FORM chunk
    const formChunk = parseFormChunk(buffer);
    console.log(`Form type: ${formChunk.formType}`);
    console.log(`Form size: ${formatSize(formChunk.size)}`);
    
    // Parse all chunks
    const chunks = parseChunks(buffer);
    console.log(`\nChunk structure:`);
    chunks.forEach((chunk, index) => {
      console.log(`${index + 1}. ${chunk.id.padEnd(4)} | Size: ${formatSize(chunk.size).padEnd(10)} | Offset: 0x${chunk.offset.toString(16).padStart(8, '0')}`);
      
      // Special handling for known chunk types
      if (chunk.id === 'FVER') {
        const version = buffer.readUInt32BE(chunk.offset + 8);
        console.log(`   Version: 0x${version.toString(16)} (${version})`);
      }
      else if (chunk.id === 'COMM') {
        const numChannels = buffer.readUInt16BE(chunk.offset + 8);
        const numFrames = buffer.readUInt32BE(chunk.offset + 10);
        const sampleSize = buffer.readUInt16BE(chunk.offset + 14);
        console.log(`   Channels: ${numChannels}, Frames: ${numFrames}, Bits: ${sampleSize}`);
        
        if (formChunk.formType === 'AIFC' && chunk.size >= 22) {
          const compType = String.fromCharCode(
            buffer[chunk.offset + 26], 
            buffer[chunk.offset + 27],
            buffer[chunk.offset + 28],
            buffer[chunk.offset + 29]
          );
          console.log(`   Compression: '${compType}'`);
        }
      }
      else if (chunk.id === 'APPL') {
        const appType = String.fromCharCode(
          buffer[chunk.offset + 8], 
          buffer[chunk.offset + 9],
          buffer[chunk.offset + 10],
          buffer[chunk.offset + 11]
        );
        
        console.log(`   App type: '${appType}'`);
        
        if (appType === 'op-1') {
          // This is OP-1 metadata, try to parse as JSON
          try {
            const jsonStart = chunk.offset + 12;
            const jsonEnd = chunk.offset + 8 + chunk.size;
            const jsonStr = buffer.slice(jsonStart, jsonEnd).toString('utf8');
            const metadata = JSON.parse(jsonStr);
            
            console.log(`   Drum version: ${metadata.drum_version}`);
            console.log(`   Name: ${metadata.name}`);
            
            if (metadata.start && metadata.start.length) {
              console.log(`   Slices: ${metadata.start.length}`);
              
              // Calculate slice spacing to check for overlap
              const sliceInfo = [];
              for (let i = 0; i < metadata.start.length; i++) {
                if (!metadata.start[i] && !metadata.end[i]) continue;
                
                const startFrames = Math.round(metadata.start[i] / 4096);
                const endFrames = Math.round(metadata.end[i] / 4096);
                const duration = (endFrames - startFrames) / 44100;
                
                sliceInfo.push({
                  index: i,
                  start: startFrames,
                  end: endFrames,
                  duration: duration.toFixed(3) 
                });
              }
              
              // Print slice information
              console.log(`\n   Slice details (first 10):`);
              for (let i = 0; i < Math.min(10, sliceInfo.length); i++) {
                const slice = sliceInfo[i];
                if (slice.start === slice.end) continue; // Skip empty slices
                console.log(`     Slice ${slice.index}: Start=${slice.start}, End=${slice.end}, Duration=${slice.duration}s`);
              }
              
              // Check for overlapping slices
              let hasOverlap = false;
              for (let i = 0; i < sliceInfo.length - 1; i++) {
                const current = sliceInfo[i];
                const next = sliceInfo[i + 1];
                if (current.end >= next.start) {
                  console.log(`   WARNING: Overlap between slices ${current.index} and ${next.index}`);
                  hasOverlap = true;
                }
              }
              
              if (!hasOverlap) {
                console.log(`   No slice overlaps detected`);
              }
            }
          } catch (e) {
            console.log(`   Error parsing metadata: ${e.message}`);
          }
        }
      }
    });
    
    return { fileType, formType: formChunk.formType, chunks };
  } catch (error) {
    console.error(`ERROR: ${error.message}`);
    return null;
  }
}

// Detect if file is AIFF or AIFC
function detectFileType(buffer) {
  // Check for FORM chunk
  if (buffer.length < 12) return 'Unknown (too small)';
  
  const formId = buffer.slice(0, 4).toString('ascii');
  if (formId !== 'FORM') return 'Not an AIFF/AIFC file';
  
  const formType = buffer.slice(8, 12).toString('ascii');
  if (formType === 'AIFF') return 'AIFF';
  if (formType === 'AIFC') return 'AIFC';
  
  return 'Unknown IFF type';
}

// Parse the main FORM chunk
function parseFormChunk(buffer) {
  const size = buffer.readUInt32BE(4);
  const formType = buffer.slice(8, 12).toString('ascii');
  
  return {
    id: 'FORM',
    size,
    formType
  };
}

// Parse all chunks in the file
function parseChunks(buffer) {
  const chunks = [];
  let pos = 12; // Skip FORM header
  
  while (pos + 8 <= buffer.length) {
    const id = buffer.slice(pos, pos + 4).toString('ascii');
    const size = buffer.readUInt32BE(pos + 4);
    
    chunks.push({
      id,
      size,
      offset: pos
    });
    
    pos += 8 + size;
    // Add pad byte if chunk size is odd
    if (size % 2 === 1) pos++;
  }
  
  return chunks;
}

// Format file size in human-readable form
function formatSize(bytes) {
  if (bytes < 1024) return `${bytes} bytes`;
  if (bytes < 1024 * 1024) return `${(bytes / 1024).toFixed(2)} KB`;
  return `${(bytes / (1024 * 1024)).toFixed(2)} MB`;
}

// Compare two files
function compareFiles(file1, file2) {
  console.log(`\n=== Comparing ${path.basename(file1)} vs ${path.basename(file2)} ===`);
  
  const data1 = analyzeAiffFile(file1);
  const data2 = analyzeAiffFile(file2);
  
  if (!data1 || !data2) {
    console.log('Cannot compare - one or both files failed analysis');
    return;
  }
  
  console.log('\n=== Key Differences ===');
  
  // Compare file types
  if (data1.fileType !== data2.fileType) {
    console.log(`File type: ${data1.fileType} vs ${data2.fileType}`);
  }
  
  if (data1.formType !== data2.formType) {
    console.log(`Form type: ${data1.formType} vs ${data2.formType}`);
  }
  
  // Compare chunk structure
  console.log('\nChunk structure:');
  
  const chunks1 = data1.chunks.map(c => c.id);
  const chunks2 = data2.chunks.map(c => c.id);
  
  console.log(`File 1: ${chunks1.join(', ')}`);
  console.log(`File 2: ${chunks2.join(', ')}`);
  
  // Find missing chunks
  const missingIn1 = chunks2.filter(c => !chunks1.includes(c));
  const missingIn2 = chunks1.filter(c => !chunks2.includes(c));
  
  if (missingIn1.length > 0) {
    console.log(`Missing in File 1: ${missingIn1.join(', ')}`);
  }
  
  if (missingIn2.length > 0) {
    console.log(`Missing in File 2: ${missingIn2.join(', ')}`);
  }
}

// Main execution
const files = process.argv.slice(2);

if (files.length === 0) {
  console.log('Usage: node analyze-aiff.js path/to/file.aif [path/to/another-file.aif]');
} else if (files.length === 1) {
  analyzeAiffFile(files[0]);
} else {
  compareFiles(files[0], files[1]);
}
